File: app/src/main/java/com/mytlogos/enterprise/viewmodel/NewsViewModel.java
package com.mytlogos.enterprise.viewmodel;

import android.app.Application;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.mytlogos.enterprise.background.TaskManager;
import com.mytlogos.enterprise.model.News;

import org.joda.time.DateTime;

import java.util.List;

public class NewsViewModel extends RepoViewModel {

    private final Handler handler;
    private final int LOADING_COMPLETE = 0x1;
    private final MutableLiveData<Boolean> loadingComplete = new MutableLiveData<>();

    public NewsViewModel(@NonNull Application application) {
        super(application);
        repository.getNews();
        handler = new Handler(Looper.getMainLooper()) {
            @Override
            public void handleMessage(Message msg) {
                loadingComplete.setValue(true);
            }
        };
    }

    public LiveData<List<News>> getNews() {
        return repository.getNews();
    }

    public void deleteOldNews() {
        repository.removeOldNews();
    }

    public LiveData<Boolean> refresh(DateTime latest) {
        TaskManager.runTask(() -> {
            try {
                repository.refreshNews(latest);
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                Message message = this.handler.obtainMessage(this.LOADING_COMPLETE);
                message.sendToTarget();
            }
        });
        return loadingComplete;
    }

    public void markNewsRead(List<Integer> readNews) {
        // TODO: 22.07.2019 send this list to client readNews
        System.out.println("read news: " + readNews);
    }
}

File: app/src/main/java/com/mytlogos/enterprise/viewmodel/NewsViewModel.kt
package com.mytlogos.enterprise.viewmodel

import android.app.Application
import android.os.Handler
import android.os.Looper
import android.os.Message
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.paging.PagedList
import com.mytlogos.enterprise.background.TaskManager.Companion.runTask
import com.mytlogos.enterprise.background.repository.NewsRepository
import com.mytlogos.enterprise.model.News
import org.joda.time.DateTime

class NewsViewModel(application: Application) : AndroidViewModel(application) {
    private val handler: Handler
    private val LOADING_COMPLETE = 0x1
    private val loadingComplete = MutableLiveData<Boolean>()
    private val newsRepository by lazy { NewsRepository.getInstance(application) }

    val news: LiveData<PagedList<News>> by lazy {
        newsRepository.news
    }

    fun deleteOldNews() {
        newsRepository.removeOldNews()
    }

    fun refresh(latest: DateTime?): LiveData<Boolean> {
        runTask {
            try {
                newsRepository.refreshNews(latest)
            } catch (e: Exception) {
                e.printStackTrace()
            } finally {
                val message = handler.obtainMessage(LOADING_COMPLETE)
                message.sendToTarget()
            }
        }
        return loadingComplete
    }

    fun markNewsRead(readNews: List<Int>) {
        // TODO: 22.07.2019 send this list to client readNews
        println("read news: $readNews")
    }

    init {
        handler = object : Handler(Looper.getMainLooper()) {
            override fun handleMessage(msg: Message) {
                loadingComplete.value = true
            }
        }
    }
}